### PR TITLE
catalog: Move LD sync from Catalog to Coordinator

### PR DESCRIPTION
**Pull request by jkosh44**
This commit moves the logic for syncing with the Launch Darkly frontend
from the Catalog to the Coordinator. This removes some unnecessary
coupling of the Catalog with the Adapter.

Works towards resolving #22593

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Review by ParkMyCar — APPROVED** [review]
Looks good to me, probably should double check with @aalexandrov too

File: src/catalog/src/durable/upgrade/stash/v44_to_v45.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    v44::ServerConfigurationKey,
+    v44::ServerConfigurationValue,
+> = TypedCollection::new("system_configuration");
+const CONFIG_COLLECTION: TypedCollection<v44::ConfigKey, v44::ConfigValue> =
+    TypedCollection::new("config");
+
+const SYSTEM_CONFIG_KEY: &str = "config_has_synced_once";
+const CONFIG_KEY: &str = "system_config_synced";
+
+/// Migrate LD sync flag from the System Configuration Collection to the Config Collection.
```
Comment (aalexandrov) [inline src/catalog/src/durable/upgrade/stash/v44_to_v45.rs:32]: Operationally it was easier keep this in the configuration collection.
Can you briefly summarize the reasons to move this before merging the PR?
Also, [the troubleshooting docs](https://www.notion.so/materialize/Troubleshooting-failing-LaunchDarkly-synchronization-006583554b714dea82880338d7375907?pvs=4#20908feae9354b309c39bb4269ed904b
) will need to be updated.

**Review by aalexandrov — APPROVED** [review]
Overall the changes look good, I'm curious why we are moving the `config_has_synced_once` value to a different collection.

**Comment by jkosh44** [comment]
> Overall the changes look good, I'm curious why we are moving the `config_has_synced_once` value to a different collection.

The main goal of this PR is to de-tangle the catalog from other parts of the adapter. This PR specifically removes the Launch darkly dependency from the catalog. So instead of syncing with launch darkly in the catalog open code, we do it in the adapter bootstrap code. At this point we do not have access to any system variables, only the config collection which is the one collection we make available to the adapter during bootstrapping. A valid take is that this refactor is too scary or adds too much of an operation burden to make it worthwhile, so feel free to push back.

As a side note an alternate approach could have been to move the launch darkly code into the catalog.

**Comment by aalexandrov** [comment]
> A valid take is that this refactor is too scary or adds too much of an operation burden to make it worthwhile, so feel free to push back.

My take is that long as we have reliable CLI tools that we can use in order to change the value of `config_has_synced_once` without bricking the catalog and we can update our existing playbooks to use these tools instead of `ALTER SYSTEM SET` this should be good to merge.

I don't know if @benesch has another take on that.

**Comment by nrainer-materialize** [comment]
Can you please take a look at the [upgrade checks?](https://buildkite.com/materialize/nightlies/builds/5526) Thanks!

**Comment by benesch** [comment]
> > A valid take is that this refactor is too scary or adds too much of an operation burden to make it worthwhile, so feel free to push back.
> 
> My take is that long as we have reliable CLI tools that we can use in order to change the value of `config_has_synced_once` without bricking the catalog and we can update our existing playbooks to use these tools instead of `ALTER SYSTEM SET` this should be good to merge. I don't know if @benesch has a different opinion, though.

Agreed!

**Comment by jkosh44** [comment]
> Can you please take a look at the [upgrade checks?](https://buildkite.com/materialize/nightlies/builds/5526) Thanks!
> 
> ```
> thread 'main' panicked at src/catalog/src/durable/upgrade.rs:293:23: upgrades not implemented
> ```

The most recent commit fixed this.

**Comment by jkosh44** [comment]
Both https://www.notion.so/materialize/Debugging-Materialize-Cloud-0788c1fe692b4acf9d247345e7b5346e#f2f12d736eae427c8892b9e561bb0be9 and https://www.notion.so/materialize/Troubleshooting-failing-LaunchDarkly-synchronization-006583554b714dea82880338d7375907#20908feae9354b309c39bb4269ed904bhave been updated.

**Pull request merged**